### ISSUE TITLE
[tune] Suggestion about Hyperband

### ISSUE THREAD

**Issue by akaniklaus** [ray-project/ray#2425]
We have been very enthusiastic about using Hyperband or PBT implementations of Raytune for hyper-parameter optimization. But we have encountered some obstacles while employing them.

- The raytune trainable function doesn't unfortunately supports the ray remote function.
- PBT can only tune the parameters that won't affect the total number of parameters.
- Hyperband and PBT can only tune less than 10 parameters, whilst TPE can tune hundreds
- Raytune has TPE interface but that doesn't support the nested definition of the search space

We see that hyperband best suits tuning categorical parameters such as the network architecture whilst PBT rather suits tuning continous numerical variables. I am curious if Raytune can enable us to bring the best of both worlds. There are several work in literature that combines Hyperband with Bayesian optimization. Hence, I would welcome if you can enlighten me about how we can do that with Raytune and would be glad if you can add that to framework as a functionality or an example.

- "Combination of Hyperband and Bayesian Optimization for Hyperparameter Optimization in Deep Learning" https://arxiv.org/pdf/1801.01596.pdf
- "Combining Hyperband and Bayesian Optimization" 
https://bayesopt.github.io/papers/2017/36.pdf 
- BOHB: Robust and Efficient Hyperparameter Optimization at Scale
https://arxiv.org/pdf/1807.01774.pdf 
- Practical Hyperparameter Optimization for Deep Learning
https://openreview.net/forum?id=HJMudFkDf

**Comment by ericl** [comment]
Thanks for the questions. Some answers inline:

> The raytune trainable function doesn't unfortunately supports the ray remote function.

Hopefully https://github.com/ray-project/ray/pull/2422 will merge soon to fix this!

> PBT can only tune the parameters that won't affect the total number of parameters.

You could create a list or dictionary of parameters and mutate them with `custom_explore_fn`.

> Raytune has TPE interface but that doesn't support the nested definition of the search space

Hm, this is not intended. Do you mean that you can't pass nested spaces in the "space" parameter for the hyperopt (TPE) scheduler here? http://ray.readthedocs.io/en/latest/tune.html#hyperopt-integration

> Integrating Hyperband with Bayesian optimization

This makes a lot of sense. Code-wise, I think the right approach here is to separate trial scheduling from trial suggestion/search, so that you can mix and match algorithms (though I'm not sure how certain combinations would work, e.g., PBT + Bayesian opt). However, HyperBand + Bayesian opt seems doable.

**Comment by richardliaw** [comment]
> Raytune has TPE interface but that doesn't support the nested definition of the search space

Can you check if the TPE interface works as intended on the latest wheels of Ray? A recent patch (#2401) may have fixed this. 

> Hyperband and PBT can only tune less than 10 parameters, whilst TPE can tune hundreds

Is this an issue with algorithmic performance or a bug?

**Comment by akaniklaus** [comment]
@ericl @richardliaw Thank you for being so super responsive. It is great to hear that #2401 and #2422 will be fixing some of those issues. I will let you know more details after trying those patches.

> Hyperband and PBT can only tune less than 10 parameters, whilst TPE can tune hundreds

This was actually a feedback given by my colleague. I don't think that it should be a performance issue because Hyperband should actually be significantly faster than TPE in an high performance computer setting. I believe that he meant that the implementations in Raytune is allowing up to 10 parameters. I will check with him again on how he concluded in that limitation of max #parameters.

Most important issue for us is either using Hyperband together with Bayesian optimization as in the papers that I have shared. They simply follow the entire process of Hyperband algorithm, but one critical difference is that instead of sampling the set of trial points independently, they sample each of them sequentially (i.e., sampling order matters) according to Bayesian optimization criterion:

> BOHB relies on HB to determine how many configurations to evaluate with which budget, but it replaces the random selection of configurations at the beginning of each HB iteration by a model-based search. We follow HB’s way of choosing the budgets and continue to use successive halving, but we replace the random sampling by a BO component to guide the search. We construct a model and use BO to select a new configuration, based on the configurations previously evaluated.

If we are able to hypertune the network topology via `custom_explore_fn` as @ericl have suggested and there is no limit on the maximum number of parameters as my co-worker assumed, I guess that we can also rather successfully employ PBT for the same purpose and we will check that again.

**Comment by richardliaw** [comment]
Hey @akaniklaus,

Just FYI, #2401 and #2422 have been merged and you can try it out now.

I've started on #2452 which addresses the important issue of enabling HyperBand with Bayesian Optimization.

**Comment by richardliaw** [comment]
Hey @akaniklaus, sorry for the late note - #2452 has been merged, so you can try it out on master now.

I'll close this for now, but feel free to reopen if you run into any other issues!

**Comment by richardliaw** [comment]
BTW, we now have BOHB - #5100. Feel free to give it a try!

**Comment by akaniklaus** [comment]
Thanks a lot!!!
